snapcraft/plugins/qbs.py

```diff
@@ -0,0 +1,157 @@
+
+This plugin uses the common plugin keywords as well as those for "sources".
+For more information check the 'plugins' topic for the former and the
+'sources' topic for the latter.
+
+Additionally, this plugin uses the following plugin-specific keywords:
+
+    - qbs-options:
+      (list of strings)
+      Additional options to pass to the qbs invocation.
+    - qt-version:
+      (enum, 'nil' or 'qt4' or 'qt5')
+      Version of Qt to use for qbs profile. If nil the base gcc/clang profile
+      will be used.
+    - qbs-profile:
+      (enum, 'gcc' or 'clang')
+      Set either gcc or clang as the base build profile. Default is gcc
+    - qbs-build-variant:
+      (enum, 'debug' or 'release')
+      Set either debug or release build variant. Default is release.
+"""
+import os
+import shutil
+import snapcraft
+
+
+class QbsPlugin(snapcraft.BasePlugin):
+
+    @classmethod
+    def schema(cls):
+        schema = super().schema()
+        schema['properties']['qbs-options'] = {
+            'type': 'array',
+            'minitems': 1,
+            'uniqueItems': True,
+            'items': {
+                'type': 'string',
+            },
+            'default': [],
+        }
+
+        schema['properties']['qbs-profile'] = {
+            'enum': ['gcc', 'clang'],
+            'default': 'gcc',
+        }
+
+        schema['properties']['qbs-build-variant'] = {
+            'enum': ['debug', 'release'],
+            'default': 'release',
+        }
+
+        schema['properties']['qt-version'] = {
+            'enum': ['nil', 'qt4', 'qt5'],
+        }
+
+        schema['required'].append('qt-version')
+
+        return schema
+
+    @classmethod
+    def get_pull_properties(cls):
+        # Inform Snapcraft of the properties associated with pulling. If these
+        # change in the YAML Snapcraft will consider the pull step dirty.
+        return ['qt-version', 'qbs-profile']
+
+    @classmethod
+    def get_build_properties(cls):
+        # Inform Snapcraft of the properties associated with building. If these
+        # change in the YAML Snapcraft will consider the build step dirty.
+        return ['qbs-options', 'qbs-build-variant']
+
+    def __init__(self, name, options, project):
+        super().__init__(name, options, project)
+        self.build_packages.extend(['qbs', 'gcc', 'clang'])
+
+        if self.options.qt_version == 'qt5':
+            self.build_packages.extend(['qt5-qmake', 'qtbase5-dev'])
+        elif self.options.qt_version == 'qt4':
+            self.build_packages.extend(['qt4-qmake', 'libqt4-dev'])
+        elif self.options.qt_version != 'nil':
+            raise RuntimeError('Unsupported Qt version: {!r}'.format(
+                self.options.qt_version))
+
+    def build(self):
+        super().build()
+
+        env = self._build_environment()
+
+        # Setup the toolchains, there will only be gcc or clang by default
+        self.run(['qbs', 'setup-toolchains', '--detect'], env=env)
```

Comment:
At this point qbs is only interested in finding toolchains it knows about and creating the base profiles for them.

The place where trouble would be caused is in the `qbs config profile.*.baseProfile = 'clang'`. Which would cause the build to fail. But really you would want that to error out at that point, as clang is expected to be in the detected profiles if clang is set in the yaml.